R/comm_manager.r

```diff
@@ -0,0 +1,186 @@
+setClassUnion('functionOrNULL', members = c('function', 'NULL'))
+
+Comm_Manager <- setRefClass(
+    'Comm_Manager',
+    fields = list(
+        send_response = 'function',
+        target_to_handler_map = 'list',
+        commid_to_comm = 'list',
+        parent_request = 'list'
+    ),
+    methods = list(
+        new_comm = function(target_name, comm_id=NULL){
+            if( is.null(comm_id) ){
+                comm_id = UUIDgenerate()
+            }
+            return (Comm$new(id=comm_id, target_name=target_name, comm_manager=.self))
+        },
+        register_target = function(target_name, handler_func) {
+            target_to_handler_map[[target_name]] <<- handler_func
+        },
+        unregister_target = function(target_name, handler_func) {
+            target_to_handler_map[[target_name]] <<- NULL
+        },
+        register_comm = function(comm){
+            commid_to_comm[[comm$id]] <<- comm
+        },
+        unregister_comm = function(comm){
+            commid_to_comm[[comm$id]] <<- NULL
+        },
+        is_comm_registered = function(comm){
+            return (!is.null(commid_to_comm[[comm$id]]))
+        },
+        send_open = function(comm_id, target_name, data, metadata=list()){
+            send_response('comm_open', parent_request, 'iopub', list(
+                metadata = metadata,
+                comm_id = comm_id,
+                target_name = target_name,
+                data = data
+            ))
+        },
+        send_msg = function(comm_id, target_name, data, metadata=list()) {
+            send_response('comm_msg', parent_request, 'iopub', list(
+                metadata = metadata,
+                comm_id = comm_id,
+                target_name = target_name,
+                data = data
+            ))
+        },
+        send_close = function(comm_id, target_name, data, metadata=list()) {
+            send_response('comm_close', parent_request, 'iopub', list(
+                metadata = metadata,
+                comm_id = comm_id,
+                target_name = target_name,
+                data = data
+            ))
+        },
+        #{
+        #   'comm_id' : 'u-u-i-d',
+        #   'target_name' : 'my_comm',
+        #   'data' : {}
+        # }
+        on_comm_open = function(request) {
+            parent_request <<- request
+            target_name <- request$content$target_name
+            comm_id <- request$content$comm_id
+
+            if(target_name %in% names(target_to_handler_map)) {
+                # create a comm object
+                comm <- new_comm(target_name, comm_id)
+                register_comm(comm)
+                data <- request$content$data
+
+                #invoke target handler
+                tryCatch({
+                    target_to_handler_map[[target_name]](comm, data)
+                }, error = function(e) {
+                    print("error invoking the handler for target")
```

Comment:
it’s OK, we’ll replace it once #119 is fixed (damn, i know that number by heart right now…)